crates/pipeline_manager/src/integration_test.rs

```diff
@@ -1084,3 +1165,105 @@ async fn primary_keys() {
+    let config = setup().await;
+    let id = deploy_pipeline_without_connectors(
+        &config,
+        r#"create table t1(id bigint not null, s varchar not null); create view v1 as select s from t1;"#,
+    )
+    .await;
+
+    // Start the pipeline
+    let resp = config
+        .post_no_body(format!("/v0/pipelines/{}/start", id))
+        .await;
+    assert_eq!(resp.status(), StatusCode::ACCEPTED);
+    config
+        .wait_for_pipeline_status(&id, PipelineStatus::Running, Duration::from_millis(1_000))
+        .await;
+
+    let mut response = config.delta_stream_request_json(&id, "V1").await;
+
+    // Push some data using default json config.
+    let req = config
+        .post_json(
+            format!(
+                "/v0/pipelines/{}/ingress/T1?format=json&update_format=insert_delete",
+                id
+            ),
+            r#"{"insert":{"id":1, "s": "1"}}
+{"insert":{"id":2, "s": "2"}}"#
+                .to_string(),
+        )
+        .await;
+    assert!(req.status().is_success());
+
+    let delta = config
+        .read_response_json(&mut response, Duration::from_millis(10_000))
+        .await
+        .unwrap();
+
+    assert_eq!(
+        delta.unwrap(),
+        json!([{"insert": {"S":"1"}}, {"insert":{"S":"2"}}])
+    );
+
+    // Push some more data
+    let req = config
+        .post_json(
+            format!(
+                "/v0/pipelines/{}/ingress/T1?format=json&update_format=insert_delete",
+                id
+            ),
+            r#"{"insert":{"id":3, "s": "3"}}
+{"insert":{"id":4, "s": "4"}}"#
+                .to_string(),
+        )
+        .await;
+    assert!(req.status().is_success());
+
+    let delta = config
+        .read_response_json(&mut response, Duration::from_millis(10_000))
+        .await
+        .unwrap();
+
+    assert_eq!(
+        delta.unwrap(),
+        json!([{"insert": {"S":"3"}}, {"insert":{"S":"4"}}])
+    );
+
+    // Push more records that will create duplicate outputs, which should be
+    // suppressed by the `outputsAreSets` SQL compiler switch.
+    let req = config
+        .post_json(
+            format!(
+                "/v0/pipelines/{}/ingress/T1?format=json&update_format=insert_delete",
+                id
+            ),
+            r#"{"insert":{"id":5, "s": "1"}}
+{"insert":{"id":6, "s": "2"}}"#
+                .to_string(),
+        )
+        .await;
+    assert!(req.status().is_success());
+
+    let delta = config
+        .read_response_json(&mut response, Duration::from_millis(10_000))
+        .await
+        .unwrap();
+
+    // Use assert_eq, so we get to see the unexpected output on error.
+    assert_eq!(delta, None);
+
+    // Shutdown the pipeline
+    let resp = config
+        .post_no_body(format!("/v0/pipelines/{}/shutdown", id))
+        .await;
+    assert_eq!(resp.status(), StatusCode::ACCEPTED);
+    config
+        .wait_for_pipeline_status(&id, PipelineStatus::Shutdown, Duration::from_millis(30_000))
```

Comment:
Use `config.shutdown_timeout` here. This value can be high in CI and k8s settings.